### PR TITLE
Fix async close race condition in RequestChannel

### LINKED ISSUE #885: AV in ServiceModel.Channels.RequestChannel+<WaitForPendingRequestsAsync>
We got an AV (null ref) in WaitForPendingRequestsAsync because _closedTcs == null.
Here is the call stack:
~~~
0:016> !clrstack
OS Thread Id: 0xfd4 (16)
 Call Site
 System.ServiceModel.Channels.RequestChannel+d__19.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.RequestChannel+d__19, System.Private.ServiceModel]](d__19 ByRef)
 System.ServiceModel.Channels.RequestChannel.WaitForPendingRequestsAsync(System.TimeSpan)
 System.ServiceModel.Channels.HttpChannelFactory`1+HttpClientRequestChannel+d__22[[System.__Canon, mscorlib]].MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.HttpChannelFactory`1+HttpClientRequestChannel+d__22[[System.__Canon, eModel]](d__22 ByRef)
 System.ServiceModel.Channels.HttpChannelFactory`1+HttpClientRequestChannel[[System.__Canon, mscorlib]].OnCloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__57.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__57, System.Private.ServiceModel]](d__57 ByRef)
 System.ServiceModel.Channels.CommunicationObject.System.ServiceModel.IAsyncCommunicationObject.CloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.ServiceChannel+d__145.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.ServiceChannel+d__145, System.Private.ServiceModel]](d__145 ByRef)
 System.ServiceModel.Channels.ServiceChannel.OnCloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__57.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__57, System.Private.ServiceModel]](d__57 ByRef)
 System.ServiceModel.Channels.CommunicationObject.System.ServiceModel.IAsyncCommunicationObject.CloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__56.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__56, System.Private.ServiceModel]](d__56 ByRef)
 System.ServiceModel.Channels.CommunicationObject.CloseAsyncInternal(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject.Close(System.TimeSpan)
 System.ServiceModel.Channels.ServiceChannelFactory.OnClose(System.TimeSpan)
 System.ServiceModel.Channels.ServiceChannelFactory+TypedServiceChannelFactory`1[[System.__Canon, mscorlib]].OnClose(System.TimeSpan)
 System.ServiceModel.Channels.ServiceChannelFactory+TypedServiceChannelFactory`1[[System.__Canon, mscorlib]].OnCloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__57.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__57, System.Private.ServiceModel]](d__57 ByRef)
 System.ServiceModel.Channels.CommunicationObject.System.ServiceModel.IAsyncCommunicationObject.CloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__56.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__56, System.Private.ServiceModel]](d__56 ByRef)
 System.ServiceModel.Channels.CommunicationObject.CloseAsyncInternal(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject.Close(System.TimeSpan)
 System.ServiceModel.ChannelFactory+d__37.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.ChannelFactory+d__37, System.Private.ServiceModel]](d__37 ByRef)
 System.ServiceModel.ChannelFactory.OnCloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__57.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__57, System.Private.ServiceModel]](d__57 ByRef)
 System.ServiceModel.Channels.CommunicationObject.System.ServiceModel.IAsyncCommunicationObject.CloseAsync(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject+d__56.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.ServiceModel.Channels.CommunicationObject+d__56, System.Private.ServiceModel]](d__56 ByRef)
 System.ServiceModel.Channels.CommunicationObject.CloseAsyncInternal(System.TimeSpan)
 System.ServiceModel.Channels.CommunicationObject.BeginClose(System.AsyncCallback, System.Object)
 System.Threading.Tasks.TaskFactory`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].FromAsyncImpl(System.Func`3, System.Func`2, System.Action`1, System.Object, onOptions)
 System.Threading.Tasks.TaskFactory.FromAsync(System.Func`3, System.Action`1, System.Object)
 *** WARNING: Unable to verify checksum for SharedPoolsOfWCFObjects.dll
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.TestHelpers.CloseFactoryAsync[[System.__Canon, mscorlib]](System.ServiceModel.ChannelFactory`1)
 SharedPoolsOfWCFObjects.CommonTest`2+c__DisplayClass21_0[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].b__0()
 SharedPoolsOfWCFObjects.CallStats+d__14.MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.CallStats.CallAsyncFuncAndRecordStatsAsync(System.Func`1, Boolean)
 SharedPoolsOfWCFObjects.CommonTest`2[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].CloseFactoryAsync(System.ServiceModel.ChannelFactory`1)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+c+b__3_1>d[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, 
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+c[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, odel.ChannelFactory`1)
 SharedPoolsOfWCFObjects.FactoryAndPoolOfItsAsyncObjects`2+d__10[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.FactoryAndPoolOfItsAsyncObjects`2[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].DestroyAsync()
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+c+b__3_4>d[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, 
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+c[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, FObjects.FactoryAndPoolOfItsAsyncObjects`2,SharedPoolsOfWCFObjects.OpenAsyncOnceChannelWrapper`1>)
 SharedPoolsOfWCFObjects.PoolOfAsyncThings`1+d__10[[System.__Canon, mscorlib]].MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.PoolOfAsyncThings`1[[System.__Canon, mscorlib]].DestroyAllPooledInstanciesImplAsync(System.__Canon[], System.Func`2)
 SharedPoolsOfWCFObjects.PoolOfAsyncThings`1+d__6[[System.__Canon, mscorlib]].MoveNext()
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.PoolOfAsyncThings`1[[System.__Canon, mscorlib]].DestoryAllPooledInstancesAsync()
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+d__8[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, 
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, Async()
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3+d__4[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, 
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[System.__Canon, mscorlib]](System.__Canon ByRef)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAndChannelsAsync_OpenOnce`3[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, ghtsAsync(Int32, Int32, Int32)
 SharedPoolsOfWCFObjects.Program+d__24`3[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]].MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAsync`3+d__7[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]].MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAsync`3+d__4[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]].MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Threading.Tasks.Task+WhenAllPromise.Invoke(System.Threading.Tasks.Task)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1)
 SharedPoolsOfWCFObjects.RecyclablePooledFactoriesAsync`3+d__5[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]].MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1)
 SharedPoolsOfWCFObjects.CallStats+d__14.MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1)
 SharedPoolsOfWCFObjects.TestHelpers+d__55`1[[System.__Canon, mscorlib]].MoveNext()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.FinishContinuations()
 System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
 System.Threading.Tasks.TaskFactory`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].FromAsyncCoreLogic(System.IAsyncResult, System.Func`2, System.Action`1, olean)
 System.Threading.Tasks.TaskFactory`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].FromAsyncCoreLogic(System.IAsyncResult, System.Func`2, System.Action`1, olean)
 System.Threading.Tasks.TaskFactory`1+c__DisplayClass35_0[[System.Threading.Tasks.VoidTaskResult, mscorlib]].b__0(System.IAsyncResult)
 System.Runtime.TaskHelpers+c.b__2_1(System.Threading.Tasks.Task, System.Object)
 System.Threading.Tasks.Task.Execute()
 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
 System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
 System.Threading.Tasks.Task.ExecuteEntry(Boolean)
 System.Threading.ThreadPoolWorkQueue.Dispatch()
~~~

### PR DESCRIPTION
**Pull request by roncain**
Stress tests revealed a race condition during an async close
of RequestChannel.  A TCS used to signal completion of the close
could become null if the final pending request completed
asynchronously during the close.

Fixes #885

### CODE REVIEW

**Comment by roncain** [comment]
@mconnew and @KKhurin -- I would appreciate your thoughts about this as the fix.  If ReleaseRequest is allowed to null the _closeTCS due to an empty list of _outstandingRequests during an async close, it defeats the WaitForPendingRequestAsync's ability to abort other pending requests if they timeout.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs [hunk 1]
```diff
@@ -196,7 +196,8 @@ private void ReleaseRequest(IRequestBase request)
                 _outstandingRequests.Remove(request);
                 if (_outstandingRequests.Count == 0)
                 {
-                    if (!_closed && _closedTcs != null)
+                    // When we are closed or closing, _closedTcs is managed by the close logic.
+                    if (!_closed && State != CommunicationState.Closing && _closedTcs != null)
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs]: You still have a race condition here, it's just a lot tighter. _closedTcs could be not null in the if expression, and then null below. I would replace the body with:
```csharp
var closedTcs = Interlocked.CompareExchange(ref _closedTcs, null, _closedTcs);
if (closedTcs != null)
{
    closedTcs.TrySetResult(null);
}
```

Comment (roncain) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs]: Good point @mconnew .  I have updated the PR and welcome any other feedback.

**Comment by roncain** [comment]
@dotnet-bot test outerloop please

**Comment by roncain** [comment]
@mconnew, @iamjasonp , @KKhurin -- can I get a signoff on this PR?  It is must-fix for RC2.
I have been running the stress test using this fixed version for the last hour with no AV, but before the fix it took < 2 minutes on my machine.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs [hunk 2]
```diff
@@ -196,10 +197,15 @@ private void ReleaseRequest(IRequestBase request)
...
+                    // When we are closed or closing, _closedTcs is managed by the close logic.
+                    if (!_closed && State != CommunicationState.Closing)
                     {
-                        _closedTcs.TrySetResult(null);
-                        _closedTcs = null;
+                        // Protect against close altering _closedTcs concurrently
+                        var closedTcs = Interlocked.CompareExchange(ref _closedTcs, null, _closedTcs);
+                        if (closedTcs != null)
+                        {
+                            closedTcs.TrySetResult(null);
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs]: Question only: there is nothing adverse happening if tcs.TrySetResult fails, right? My read is that this returns false if the tcs is already in a final state, so just wondering if this meant something weird could have happened

**Comment by iamjasonp** [comment]
LGTM. Would prefer if @mconnew or @KKhurin could also take a look as well, but if time is a factor I think your stress test is proof enough to me that we are ok.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs (hunk 2 above)
Comment (roncain) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs]: My read is that if the close succeeded in its own TrySetResult (and TCS was not yet null -- which should not be possible given the state checks above), the TrySetResult here would be a NOP.  And nothing checks the return code for TrySetResult.  Translation: I think we're good.

**Comment by roncain** [comment]
Time is a factor.  The RC2 release branch will be created any time now, and not checking in before then will require a trip through ship room.  I've tried OuterLoops on 2 machines and run the stress test for now around 90 minutes without an AV.  I'll merge to make the RC branch cut-off and accept post-merge feedback.

**Pull request merged**